spyder/app/mainwindow.py

```diff
@@ -237,7 +237,6 @@ def register_plugin(self, plugin_name, external=False, omit_conf=False):
         plugin.sig_exception_occurred.connect(self.handle_exception)
         plugin.sig_free_memory_requested.connect(self.free_memory)
         plugin.sig_quit_requested.connect(self.close)
-        plugin.sig_restart_requested.connect(self.restart)

```

Comment:
I guess I'm confused. The first option effectively makes the mainwindow a central dispatcher, so other plugins don't access the application plugin directly. I thought we were moving away from that.
Also, the second option doesn't require calling `application.restart` or connecting another slot to `sig_restart_requested`. If the plugin wants to request a restart, it just needs to emit the signal.

Perhaps I'm still confused about the architecture. My understanding is as follows:
1. The application plugin owns the `restart` slot and the `sig_restart_requested` signal. They are connected.
2. Any plugin that wants to _listen_ for `sig_restart_requested`, must require the application plugin and connect the signal to whatever slot it wants executed, e.g. shut something down. `application.restart` doesn't need to be connected here, since that is already connected to the signal in the application plugin and will be executed when the signal is emitted (no matter which plugin emits it).
3. Any plugin that wants to _request_ a restart must require the application plugin and simply emit `application.sig_restart_requested`. It should not call `application.restart`.

Reply:
Ok, sorry for introducing more confusion myself. This comment helped me to understand your rationale behind this PR and I think it's pretty good.

Please ignore all my comments about this and proceed with your initial plan.